File: modules/meeterm-terminal/android/src/main/java/dev/meeterm/terminal/TerminalRegistry.kt
package dev.meeterm.terminal

/**
 * Process-local mapping from the durable terminal identity to its Rust handle.
 *
 * A view may be recreated while React Native changes screens or surfaces. A
 * release only drops the view reference; it intentionally does not destroy
 * the Rust terminal. Destruction is reserved for an explicit test reset or a
 * future process-shutdown owner.
 */
internal object TerminalRegistry {
  private data class Entry(
    val handle: Long,
    var references: Int,
  )

  private val entries = mutableMapOf<String, Entry>()

  @Synchronized
  fun acquire(terminalId: String, columns: Int, rows: Int): Long {
    val handle = ensure(terminalId, columns, rows)
    entries[terminalId]?.references = (entries[terminalId]?.references ?: 0) + 1
    return handle
  }

  /**
   * Return the stable handle without taking a view reference. Control-plane
   * calls can arrive before the first native view has attached, so they use
   * this method to create the Rust terminal while leaving view lifetime
   * accounting independent from the SSH connection lifetime.
   */
  @Synchronized
  fun ensure(terminalId: String, columns: Int, rows: Int): Long {
    // Pane handles already belong to the one Rust registry. Never allocate a
    // second terminal when a view switches to a remote pane.
    if (terminalId.startsWith("native:")) return nativeHandle(terminalId)
    entries[terminalId]?.let { return it.handle }

    val handle = MeetermNative.create(columns, rows)
    if (handle == 0L) return 0L
    entries[terminalId] = Entry(handle, references = 0)
    return handle
  }

  @Synchronized
  fun release(terminalId: String, handle: Long) {
    val entry = entries[terminalId] ?: return
    if (entry.handle != handle) return
    entry.references = (entry.references - 1).coerceAtLeast(0)
  }

  @Synchronized
  fun handleFor(terminalId: String): Long =
    if (terminalId.startsWith("native:")) nativeHandle(terminalId)
    else entries[terminalId]?.handle ?: 0L

  private fun nativeHandle(terminalId: String): Long {
    val handle = terminalId.removePrefix("native:").toLongOrNull()
      ?.takeIf { it > 0L } ?: return 0L
    return if (MeetermNative.terminalExists(handle)) handle else 0L
  }

  /** Explicit test/process reset. Normal view unmount does not call this. */
  @Synchronized
  fun resetForTests() {
    entries.values.forEach { entry ->
      MeetermNative.destroy(entry.handle)
    }
    entries.clear()
  }
}

File: modules/meeterm-terminal/ios/TerminalRegistry.swift
import Foundation

/// Process-local stable identity mapping. Views borrow a handle; unmounting a
/// view never destroys its Rust terminal state.
final class TerminalRegistry {
  private static let lock = NSLock()
  private static var handles: [String: UInt64] = [:]

  static func acquire(terminalId: String, columns: Int, rows: Int) -> UInt64 {
    lock.lock()
    defer { lock.unlock() }

    return ensureLocked(terminalId: terminalId, columns: columns, rows: rows)
  }

  /// Resolve a stable terminal handle for control-plane calls that can arrive
  /// before a TerminalView has attached. This does not make view lifecycle the
  /// owner of the Rust terminal and therefore does not disconnect on unmount.
  static func ensure(terminalId: String, columns: Int, rows: Int) -> UInt64 {
    lock.lock()
    defer { lock.unlock() }
    return ensureLocked(terminalId: terminalId, columns: columns, rows: rows)
  }

  static func handle(for terminalId: String) -> UInt64 {
    lock.lock()
    defer { lock.unlock() }
    if terminalId.hasPrefix("native:") { return nativeHandle(terminalId) }
    return handles[terminalId] ?? 0
  }

  private static func ensureLocked(terminalId: String, columns: Int, rows: Int) -> UInt64 {
    // Rust owns pane handles and their lifetime; a view only borrows them.
    if terminalId.hasPrefix("native:") { return nativeHandle(terminalId) }
    if let existing = handles[terminalId] {
      return existing
    }

    let handle = MeetermCore.create(columns: columns, rows: rows)
    guard handle != 0 else {
      return 0
    }
    handles[terminalId] = handle
    return handle
  }

  private static func nativeHandle(_ terminalId: String) -> UInt64 {
    guard let handle = UInt64(terminalId.dropFirst("native:".count)),
          handle != 0, MeetermCore.terminalExists(terminalId: handle) else { return 0 }
    return handle
  }

  /// Destruction is explicit and reserved for tests/process ownership. Normal
  /// Expo view lifecycle changes intentionally do not call this function.
  static func resetForTests() {
    lock.lock()
    defer { lock.unlock() }

    for handle in handles.values {
      MeetermCore.destroy(terminalId: handle)
    }
    handles.removeAll(keepingCapacity: false)
  }
}
